File: 剑指 Offer（第 2 版）/28.cpp
// 剑指 Offer 28. 对称的二叉树
// https://leetcode-cn.com/problems/dui-cheng-de-er-cha-shu-lcof/

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(!root) return true;
        return recur(root->left, root->right);
    }

    bool recur(TreeNode* left, TreeNode* right){
        if((!left)&&(!right)) return true;
        if((!left)||(!right)||left->val!=right->val) return false;
        return recur(left->left, right->right) && recur(left->right, right->left);
    }
};

File: 剑指 Offer（第 2 版）/28.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    public boolean isSymmetric(TreeNode root) {
        if(root==null) return true;
        return recur(root.left, root.right);
    }
    public boolean recur(TreeNode left, TreeNode right){
        if(left==null&&right==null) return true;
        if(left==null||right==null||left.val!=right.val) return false;
        // 传进去的是需要对称的两棵小树的根
        return recur(left.left, right.right) && recur(left.right, right.left);
    }
}